sklearn/decomposition/tests/test_dict_learning.py

```diff
@@ -1024,3 +1025,60 @@ def test_minibatch_dictionary_learning_warns_and_ignore_n_iter():
     with pytest.warns(FutureWarning, match=warn_msg):
         model = MiniBatchDictionaryLearning(batch_size=256, n_iter=2, max_iter=2).fit(X)
     assert model.n_iter_ == 2
+
+
+class SparseCoderForCommonTests(SparseCoder):
+    """Inheriting from SparseCoder to pass the common test.
+
+    SparseCoder is fundamentally stateless and requires a dictionary parameter
+    as constructor parameter which makes it challenging to run the usual transformer
+    checks. This stateful subclass instead creates a random dictionary that is
+    compatible with the input data dimensions passed to fit to benefit from the
+    test coverage provided by the usual stateful transformer checks.
+    """
```

Comment:
O_o, interesting. When we work on allowing custom data generators for tests, this would then probably be void.

Reply:
Actually, we should not have that. I think that now that we have metadata routing, we should instead pass `dictionary` as a metadata and not as an `__init__` parameter. This will avoid this subclassing.

However, we need to have something that generate the necessary metadata thought.

Reply:
I'd be happy if we deprecate it on the constructor and move it to the relevant method here as well.